scripts/collect_steps.py

```diff
@@ -0,0 +1,165 @@
+        bufsize=1,
+        universal_newlines=True,
+    )
+
+    count = 0
+    header = 0
+    while True:
+        line = proc.stdout.readline()
+        if not line or line == "":
+            break
+
+        if not line.startswith(ROOT_STEP_PREFIX):
+            header += 1
+            continue
+
+        output.append(remove_root_step_name_from_line(line))
+        count += 1
+
+    # safeguard against empty output
+    if count == 0:
+        print("No steps in the output")
+        exit(1)
+
+    # oneshot_ipa produces 4 header lines
+    if header != EXPECTED_HEADER_LINES_COUNT:
+        print("Unexpected header lines in the output")
+        exit(1)
+
+    return output
+
+
+# Splits a line by "/" and create a vector consisting each splitted string
+# concatenated by all the preceding strings.
+#
+# # Example
+# input = "mod_conv_match_key/mc0/mc0/xor1"
+# output = ["mod_conv_match_key",
+#           "mod_conv_match_key/mc0",
+#           "mod_conv_match_key/mc0/mc0",
+#           "mod_conv_match_key/mc0/mc0/xor1"]
+#
+# We do this because not all substeps invoke communications between helpers.
+#
+# For example, a leaf substep "mod_conv_match_key/mc0/mc0/xor1" invokes
+# multiplications, but "mod_conv_match_key/mc0/mc0" and "mod_conv_match_key/mc0"
+# do not. However, we want to include these intermediate substeps in the output
+# since each `narrow`, even if it doesn't actually do anything, is a state
+# transition.
+#
+# This function generates a lot of duplicates. It is inefficient but we don't
+# really care because we execute this script only once when we add a new
+# protocol which is a pretty rare case. The duplicates will be removed in the
+# `remove_duplicates_and_sort`.
+def extract_intermediate_steps(steps):
+    output = []
+    for step in steps:
+        substeps = step.split("/")
+        for i in range(1, len(substeps)):
+            output.append("/".join(substeps[:i]))
+    steps.extend(output)
+
+
+def remove_duplicates_and_sort(steps):
+    # Converting a list to dict removes duplicates. We want a list for the sorting.
+    unique_steps = list(dict.fromkeys(steps))
+
+    # "natural" sort
+    unique_steps.sort(key=natural_sort_key)
+
+    # remove empty string if present
+    try:
+        unique_steps.remove("")
+    except Exception:
+        pass
+
+    return unique_steps
+
+
+if __name__ == "__main__":
+    steps = []
+    for c in PER_USER_CAP:
+        for w in ATTRIBUTION_WINDOW:
+            for b in BREAKDOWN_KEYS:
+                for m in SECURITY_MODEL:
+                    args = ARGS + [
+                        "-n",
+                        str(QUERY_SIZE),
+                        "-c",
+                        str(c),
+                        "-w",
+                        str(w),
+                        "-b",
+                        str(b),
+                        "-m",
+                        m,
+                    ]
+                    steps.extend(collect_steps(args))
```

Comment:
keeping a set might be more efficient as you don't need space for all the duplicates

Reply:
Changed to use sets.